tools/src/merger/OpenApiVersionExtractor.ts

```diff
@@ -0,0 +1,104 @@
+/*
+* Copyright OpenSearch Contributors
+* SPDX-License-Identifier: Apache-2.0
+*
+* The OpenSearch Contributors require contributions made to
+* this file be licensed under the Apache-2.0 license or a
+* compatible open source license.
+*/
+
+import _, { extend, isEmpty } from 'lodash'
+import { delete_matching_keys, find_refs, write_yaml } from '../helpers'
+import { Logger } from '../Logger'
+import { type OpenAPIV3 } from 'openapi-types'
+import semver from 'semver'
+
+// Extract a versioned API
+export default class OpenApiVersionExtractor {
+  private _spec?: Record<string, any>
+  private _source_spec: OpenAPIV3.Document
+  private _target_version?: string
+  private _logger: Logger
+
+  constructor(source_spec: OpenAPIV3.Document, target_version?: string, logger: Logger = new Logger()) {
+    this._source_spec = source_spec
+    this._target_version = target_version !== undefined ? semver.coerce(target_version)?.toString() : undefined
+    this._logger = logger
+    this._spec = undefined
+  }
+
+  extract(): OpenAPIV3.Document {
+    if (this._spec) return this._spec as OpenAPIV3.Document
+    if (this._target_version !== undefined) {
+      this.#extract()
+    } else {
+      this._spec = this._source_spec
+    }
+    return this._spec as OpenAPIV3.Document
+  }
+
+  write_to(output_path: string): OpenApiVersionExtractor {
+    this._logger.info(`Writing ${output_path} ...`)
+    write_yaml(output_path, this.extract())
+    return this
+  }
+
+  // Remove any refs and objects that reference them that are x-version-added/removed
+  // incompatible with the target server version.
+  #extract() : void {
+    this._logger.info(`Extracting version ${this._target_version} ...`)
+
+    this._spec = _.cloneDeep(this._source_spec)
+
+    this.#remove_keys_not_matching_semver()
+    this.#remove_unused()
+  }
+
+  #exclude_per_semver(obj: any): boolean {
+    if (this._target_version === undefined) return false
+
+    const x_version_added = semver.coerce(obj['x-version-added'] as string)
+    const x_version_removed = semver.coerce(obj['x-version-removed'] as string)
+
+    if (x_version_added && !semver.satisfies(this._target_version, `>=${x_version_added?.toString()}`)) {
+      return true
+    } else if (x_version_removed && !semver.satisfies(this._target_version, `<${x_version_removed?.toString()}`)) {
+      return true
+    }
+
+    return false
+  }
+
+  // Remove any elements that are x-version-added/removed incompatible with the target server version.
+  #remove_keys_not_matching_semver(): void {
+    if (this._target_version === undefined) return
+    delete_matching_keys(this._spec, this.#exclude_per_semver.bind(this))
+  }
+
+  #remove_unused(): void {
+    if (this._spec === undefined) return
+
+    // remove anything that's not referenced
+    var references: string[] = find_refs(this._spec)
```

Comment:
`references`, in this case, will also include all indirectly referenced components whose grandparents have been removed (since `find_refs` will traverse `_spec.components` as well). That is, say `#/components/parameters/X` has been removed. but `X` is an array whose `items` is `$ref: #/components/schemas/Y` but `Y` is `anyOf: [A,B]`. In this case, after X is removed, Y will be removed by this function as a result, but A and B will remain.

On the other hand, `find_refs(this.spec.paths)` will exclude all `#components/schemas/...` references being used by other `#/components/<non-schemas>/...` 😅. That's not good either.

I'm afraid this strategy won't work. You'll have to follow (1) described [here](https://github.com/opensearch-project/opensearch-api-specification/pull/428#issuecomment-2243844698) to find all `$ref` actually being used by the remaining parts of `#paths` (directly or indirectly).

Reply:
https://github.com/opensearch-project/opensearch-api-specification/pull/428/commits/ed18c4d145a39c8a3c3cf2894dcfa66175102df8 will traverse refs recursively.

Reply:
resolved

Reply:
> On the other hand, find_refs(this.spec.paths) will exclude all #components/schemas/... references being used by other #/components/<non-schemas>/... 😅. That's not good either.

This is now handled. If a component is referenced from within `paths`, all its refs are now safe.